File: src/cryptotax/api/schemas/transactions.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class TransactionResponse(BaseModel):
    id: int
    wallet_id: uuid.UUID
    chain: str
    tx_hash: str
    block_number: Optional[int]
    timestamp: Optional[int]
    from_addr: Optional[str]
    to_addr: Optional[str]
    value_wei: Optional[int]
    gas_used: Optional[int]
    status: str
    entry_type: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class TransactionDetail(TransactionResponse):
    tx_data: Optional[str]
    updated_at: datetime


class TransactionList(BaseModel):
    transactions: list[TransactionResponse]
    total: int
    limit: int
    offset: int

File: web/src/api/transactions.ts
import { apiFetch, withEntityId } from './client'

export interface Transaction {
  id: number
  wallet_id: string
  chain: string
  tx_hash: string
  block_number: number | null
  timestamp: number | null
  from_addr: string | null
  to_addr: string | null
  value_wei: number | null
  gas_used: number | null
  status: string
  entry_type: string | null
  created_at: string
}

export interface TransactionDetail extends Transaction {
  tx_data: string | null
  updated_at: string
}

export interface TransactionList {
  transactions: Transaction[]
  total: number
  limit: number
  offset: number
}

export interface TransactionFilters {
  wallet_id?: string
  chain?: string
  status?: string
  limit?: number
  offset?: number
  date_from?: string
  date_to?: string
}

export async function listTransactions(filters: TransactionFilters = {}, entityId?: string): Promise<TransactionList> {
  const params = new URLSearchParams()
  if (filters.wallet_id) params.set('wallet_id', filters.wallet_id)
  if (filters.chain) params.set('chain', filters.chain)
  if (filters.status) params.set('status', filters.status)
  if (filters.limit) params.set('limit', String(filters.limit))
  if (filters.offset !== undefined) params.set('offset', String(filters.offset))
  if (filters.date_from) params.set('date_from', filters.date_from)
  if (filters.date_to) params.set('date_to', filters.date_to)

  const qs = params.toString()
  const path = `/transactions${qs ? `?${qs}` : ''}`
  return apiFetch<TransactionList>(withEntityId(path, entityId))
}

export async function getTransaction(txHash: string): Promise<TransactionDetail> {
  return apiFetch<TransactionDetail>(`/transactions/${txHash}`)
}
